domains/client/domain-operator/src/bundle_processor.rs

```diff
@@ -72,30 +80,155 @@ where
     }
 }
 
-struct ReceiptValidator<Client> {
+// TODO: Remove PhantomData once fraud proof generator and parent chain are added
+struct ReceiptValidator<Client, Block, CBlock> {
+    domain_id: DomainId,
     client: Arc<Client>,
+    _phantom_block: PhantomData<Block>,
+    _phantom_cblock: PhantomData<CBlock>,
 }
 
-impl<Client> Clone for ReceiptValidator<Client> {
+impl<Client, Block, CBlock> Clone for ReceiptValidator<Client, Block, CBlock> {
     fn clone(&self) -> Self {
         Self {
+            domain_id: self.domain_id,
             client: self.client.clone(),
+            _phantom_block: self._phantom_block,
+            _phantom_cblock: self._phantom_cblock,
+        }
+    }
+}
+
+impl<Client, Block, CBlock> ReceiptValidator<Client, Block, CBlock> {
+    pub fn new(domain_id: DomainId, client: Arc<Client>) -> Self {
+        Self {
+            domain_id,
+            client,
+            _phantom_block: PhantomData,
+            _phantom_cblock: PhantomData,
+        }
+    }
+}
+
+/// Verifies invalid_bundle field in the ER and generates fraud proof in case the field
+/// is incorrect. Fraud proof refers to the first mismatch.
+/// CONTRACT: It will return None if the field is valid, otherwise it will return `Some` with fraud proof
+/// pointing to first mismatch in invalid bundles array.
+pub fn verify_and_generate_fraud_proof_for_invalid_bundles<Block, CBlock>(
```

Comment:
How about separating the fraud detection and fraud proof generation into 2 func:
```rust
fn verify_invalid_bundles(...) -> Option<ReceiptMismatchInfo> { ... }

fn generate_invalid_bundles_fraud_proof(mismatch: ReceiptMismatchInfo) -> InvalidBundleFraudProof { ... }
```

And similar to the invalid reward info fraud proof, these 2 func should be invoked in:
https://github.com/subspace/subspace/blob/fd63aa4831cc6669f489d2344bff496390a8833a/domains/client/domain-operator/src/domain_block_processor.rs#L691
https://github.com/subspace/subspace/blob/fd63aa4831cc6669f489d2344bff496390a8833a/domains/client/domain-operator/src/domain_block_processor.rs#L797

Reply:
Only problem being that if `generate_invalid_bundles_fraud_proof` accepts `ReceiptMismatchInfo` from the signature alone, it is expected to convert all variant of `ReceiptMismatchInfo` enum to fraud proof.  Which is right now handled primarily by `ReceiptChecker` (which from my understanding is deprecated.)

In this case it is better to leave this as it is, and do the refactoring where we transfer the handling of `ReceiptMismatchInfo` from `ReceiptChecker` to `ReceiptValidator`. What do you think?

Reply:
The above is just for demonstration purposes, if the interface doesn't fit we can certainly change it to something like:
```rust
fn generate_invalid_bundles_fraud_proof(x, y, z) -> InvalidBundleFraudProof { ... }
...
ReceiptMismatchInfo::InvalidBundlesMismatch { x, y, z } => self.generate_invalid_bundles_fraud_proof(x, y, z)
```
plz take a look at the above linked code for how the invalid reward fraud proof is handled.

As for `ReceiptValidator`, we may not need it anymore because we won't consider the bundle with the bad receipt as an invalid bundle anymore (see https://github.com/subspace/subspace/pull/1957#discussion_r1326123748), so we don't need to check the receipt at `block-preprocessor` and it is actually okay to delete `ReceiptValidator`.

Reply:
Okay. We can remove `ReceiptValidator` in next PR as it is out of scope for this PR. For now, I migrated changes to the `ReceiptsChecker`.